runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/PipelineNode.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction.graph;
+
+import com.google.auto.value.AutoValue;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PCollection;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PTransform;
+
+/**
+ * A graph node which contains some pipeline element.
+ */
+interface PipelineNode {
+  enum NodeType {
+    PCOLLECTION,
+    PTRANSFORM
+  }
+
+  static PTransformNode ptransform(String id, PTransform transform) {
+    return new AutoValue_PipelineNode_PTransformNode(id, transform);
+  }
+
+  static PCollectionNode pcollection(String id, PCollection collection) {
+    return new AutoValue_PipelineNode_PCollectionNode(id, collection);
+  }
+
+  NodeType getType();
```

Comment:
Is this so you have a sane switch statement?

It won't be effective unless you have groups of node types because as you transform the graph over time you'll find that your inserting nodes that contain more runtime information.

Reply:
rm'd this; instead I'm using the bipartite nature of the graph as appropriate.